File: apps/cli/src/commands/suggest.ts
import { readFileSync } from "node:fs";
import { basename } from "node:path";

import { Command } from "commander";

import { formatBold, formatDim, formatType } from "../lib/colors";
import { output, outputError } from "../lib/output";
import { getServerUrl } from "../lib/store";

interface ExistingChunk {
    id: string;
    title: string;
    type: string;
    content: string;
}

interface Suggestion {
    type: string;
    title: string;
    reason: string;
}

export const suggestCommand = new Command("suggest")
    .description("Suggest chunks to create for a file")
    .argument("<path>", "source file path")
    .option("--codebase <name>", "scope to codebase")
    .action(async (filePath: string, opts: { codebase?: string }, cmd: Command) => {
        let content: string;
        try {
            content = readFileSync(filePath, "utf-8");
        } catch {
            outputError(`Cannot read file: ${filePath}`);
            process.exit(1);
        }

        const lines = content.split("\n");

        // Analyze complexity
        const stats = {
            lines: lines.length,
            functions: (content.match(/(?:function |const \w+ = (?:async )?\(|(?:export )?(?:async )?function)/g) || []).length,
            imports: (content.match(/^import /gm) || []).length,
            exports: (content.match(/^export /gm) || []).length,
            classes: (content.match(/^(?:export )?class /gm) || []).length,
        };

        // Check existing coverage
        let serverUrl: string | undefined;
        try {
            serverUrl = getServerUrl();
        } catch {
            outputError('No server URL configured. Run "fubbik init" first.');
            process.exit(1);
        }

        if (!serverUrl) {
            outputError('No server URL configured. Run "fubbik init" first.');
            process.exit(1);
        }

        let existingChunks: ExistingChunk[] = [];
        try {
            const params = new URLSearchParams({ path: filePath });
            if (opts.codebase) {
                params.set("codebaseId", opts.codebase);
            }
            const res = await fetch(`${serverUrl}/api/context/for-file?${params.toString()}`);
            if (res.ok) {
                existingChunks = (await res.json()) as ExistingChunk[];
            }
        } catch {
            // Continue without existing chunks if server is unreachable
        }

        // Suggest based on gaps
        const suggestions: Suggestion[] = [];
        const fileName = basename(filePath);

        if (existingChunks.length === 0) {
            suggestions.push({
                type: "reference",
                title: `${fileName} Documentation`,
                reason: "No chunks reference this file",
            });
        }

        if (stats.functions > 5 && !existingChunks.some(c => c.type === "reference")) {
            suggestions.push({
                type: "reference",
                title: `${fileName} API Reference`,
                reason: `${stats.functions} functions found but no reference chunk`,
            });
        }

        if (stats.exports > 3) {
            suggestions.push({
                type: "document",
                title: `${fileName} Architecture`,
                reason: `${stats.exports} exports suggest this is a key module`,
            });
        }

        if (stats.lines > 200) {
            suggestions.push({
                type: "note",
                title: `${fileName} Conventions`,
                reason: `Large file (${stats.lines} lines) may have implicit conventions`,
            });
        }

        if (stats.classes > 0 && !existingChunks.some(c => c.type === "schema")) {
            suggestions.push({
                type: "schema",
                title: `${fileName} Class Schema`,
                reason: `${stats.classes} class(es) found but no schema chunk`,
            });
        }

        // Output
        if (suggestions.length === 0) {
            const humanText = formatDim("No suggestions — file appears well-covered");
            output(cmd, { suggestions: [], stats }, humanText);
        } else {
            const humanLines: string[] = [
                formatBold(`${suggestions.length} suggestion(s) for ${filePath}:`),
                formatDim(`  Analysis: ${stats.lines} lines, ${stats.functions} functions, ${stats.imports} imports, ${stats.exports} exports`),
                "",
            ];
            for (const s of suggestions) {
                humanLines.push(`  ${formatType(s.type)} ${formatBold(s.title)}`);
                humanLines.push(`    ${formatDim(s.reason)}`);
            }
            output(cmd, { suggestions, stats }, humanLines.join("\n"));
        }
    });

File: crates/fubbik-cli/src/commands/suggest.rs
use std::path::Path;

use anyhow::{Context, Result};

use crate::client::Client;
use crate::output::{self, OutputMode};

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
struct FileStats {
    lines: usize,
    functions: usize,
    imports: usize,
    exports: usize,
    classes: usize,
}

#[derive(Debug, PartialEq, Eq, serde::Serialize)]
struct Suggestion {
    #[serde(rename = "type")]
    chunk_type: String,
    title: String,
    reason: String,
}

pub async fn run(
    client: &Client,
    path: &Path,
    space: Option<&str>,
    mode: OutputMode,
) -> Result<()> {
    let content = std::fs::read_to_string(path)
        .with_context(|| format!("cannot read file: {}", path.display()))?;
    let stats = analyze(&content);
    let space = client.resolve_space(space).await?;
    let context = client
        .context_for_file(
            &path.to_string_lossy(),
            space.as_deref(),
            8000,
            "json-legacy",
        )
        .await?;
    let existing = context["chunks"].as_array().cloned().unwrap_or_default();
    let file_name = path
        .file_name()
        .and_then(|name| name.to_str())
        .unwrap_or("file");
    let suggestions = suggestions(file_name, &stats, &existing);
    let result = serde_json::json!({ "suggestions": suggestions, "stats": stats });
    if mode == OutputMode::Json {
        return output::json(&result);
    }
    if suggestions.is_empty() {
        println!("No suggestions — file appears well-covered");
        return Ok(());
    }
    println!(
        "{} suggestion(s) for {}:",
        suggestions.len(),
        path.display()
    );
    println!(
        "  Analysis: {} lines, {} functions, {} imports, {} exports\n",
        stats.lines, stats.functions, stats.imports, stats.exports
    );
    for suggestion in suggestions {
        println!(
            "  [{}] {}\n    {}",
            suggestion.chunk_type, suggestion.title, suggestion.reason
        );
    }
    Ok(())
}

fn analyze(content: &str) -> FileStats {
    let lines = content.split('\n').count();
    let imports = content
        .lines()
        .filter(|line| line.starts_with("import "))
        .count();
    let exports = content
        .lines()
        .filter(|line| line.starts_with("export "))
        .count();
    let classes = content
        .lines()
        .filter(|line| {
            let line = line.trim_start();
            line.starts_with("class ") || line.starts_with("export class ")
        })
        .count();
    let functions = content.match_indices("function ").count()
        + content.match_indices(" = (").count()
        + content.match_indices(" = async (").count();
    FileStats {
        lines,
        functions,
        imports,
        exports,
        classes,
    }
}

fn suggestions(
    file_name: &str,
    stats: &FileStats,
    existing: &[serde_json::Value],
) -> Vec<Suggestion> {
    let has_type = |kind: &str| existing.iter().any(|chunk| chunk["type"] == kind);
    let mut result = Vec::new();
    if existing.is_empty() {
        result.push(suggestion(
            "reference",
            format!("{file_name} Documentation"),
            "No chunks reference this file".into(),
        ));
    }
    if stats.functions > 5 && !has_type("reference") {
        result.push(suggestion(
            "reference",
            format!("{file_name} API Reference"),
            format!("{} functions found but no reference chunk", stats.functions),
        ));
    }
    if stats.exports > 3 {
        result.push(suggestion(
            "document",
            format!("{file_name} Architecture"),
            format!("{} exports suggest this is a key module", stats.exports),
        ));
    }
    if stats.lines > 200 {
        result.push(suggestion(
            "note",
            format!("{file_name} Conventions"),
            format!(
                "Large file ({} lines) may have implicit conventions",
                stats.lines
            ),
        ));
    }
    if stats.classes > 0 && !has_type("schema") {
        result.push(suggestion(
            "schema",
            format!("{file_name} Class Schema"),
            format!("{} class(es) found but no schema chunk", stats.classes),
        ));
    }
    result
}

fn suggestion(chunk_type: &str, title: String, reason: String) -> Suggestion {
    Suggestion {
        chunk_type: chunk_type.into(),
        title,
        reason,
    }
}

#[cfg(test)]
mod tests {
    use super::{analyze, suggestions};

    #[test]
    fn uncovered_complex_file_produces_reference_architecture_and_schema_suggestions() {
        // Given an uncovered source file with many functions, exports, and a class
        let source = "export class Api {}\nexport function a() {}\nexport function b() {}\nexport function c() {}\nexport function d() {}\nexport function e() {}\nexport function f() {}";
        let stats = analyze(source);

        // When suggestions are generated without existing chunks
        let suggestions = suggestions("api.ts", &stats, &[]);

        // Then documentation, API, architecture, and schema gaps are reported
        assert_eq!(suggestions.len(), 4);
        assert_eq!(suggestions[0].chunk_type, "reference");
        assert_eq!(suggestions[1].title, "api.ts API Reference");
        assert_eq!(suggestions[2].chunk_type, "document");
        assert_eq!(suggestions[3].chunk_type, "schema");
    }
}
